service/src/main/kotlin/fi/espoo/evaka/daycare/service/AbsenceService.kt

```diff
@@ -90,6 +93,54 @@ class AbsenceService {
     }
 }
 
+private fun supplementReservationsWithDailyServiceTimes(
+    placementDateRanges: List<FiniteDateRange>,
+    unitOperationalDays: Set<LocalDate>,
+    reservations: List<ChildReservation>?,
+    dailyServiceTimes: DailyServiceTimes?,
+    absences: List<AbsenceWithModifierInfo>?
+): List<HelsinkiDateTimeRange> {
+    val reservationRanges = reservations
+        ?.filter { reservation -> absences?.none { it.date == reservation.startTime.toLocalDate() } ?: true }
+        ?.map { HelsinkiDateTimeRange(it.startTime, it.endTime) }
+        ?: listOf()
+
+    val dailyServiceTimeRanges =
+        if (dailyServiceTimes == null) listOf()
+        else placementDateRanges
+            .flatMap { it.dates() }
+            .filter { date -> reservationRanges.none { it.start.toLocalDate() == date } }
```

Comment:
This might be slightly more readable if reservation dates were mapped to LocalDates out of lambda's scope.

```kt
val reservationDates  = reservationRanges.map { it.start.toLocalDate() }
[..]
.filterNot { date -> reservationDates.contains(date) }
```

Furthermore, are reservations always on the same day or can they span multiple days?

Reply:
Reservations can continue into the next day, but they are always at most 24h long.

Reply:
Should these filtering conditions take end date of reservation somehow into account too?

Reply:
Eg. If the reservation is `23:00 - 14:00`, only the first day gets dropped here. Is it fine?

Reply:
They intentionally don't because otherwise it would be impossible to mark that the child is absent the next evening.

Reply:
I'll add a couple of tests.